### PR TITLE
Upgrade Hibernate and run unit tests with PG 11

### PR DESCRIPTION
**Pull request by coverbeck**
https://ucsc-cgl.atlassian.net/browse/SEAB-2676

Change unit tests to use PG 11, which opened a whole can of worms. Lots of details in [SEAB-2676](https://ucsc-cgl.atlassian.net/browse/SEAB-2676). Short version here:

1. Postgres 10 introduced support for SQL Identity.
2. Pre PG-10, Liquibase/Hibernate would simulate SQL Identity with SQL Sequence.
3. When we upgraded from PG 9 to PG11, Liquibase took advantage of SQL Identity, but Hibernate didn't, leading to the propose_migrations failures (Liquibase would generate identity, Hibernate would generate Sequence)
4. We need to tell Hibernate to use the PG10 dialect in the dockstore.yml, so that it will also generate Identity, but our version of Hibernate doesn't have that dialect.

So, upgrade Hibernate to a version that has the PostgreSQL10 dialect. The version bundled with DropWizard (we were behind), supports it. So upgrading it just means not pinning to an older version. And then chasing down the enforcer plugin errors.

Note that if you run the Liquibase migrations from scratch on PG 9 vs PG 11, you will end up with different schemas; there will be identity vs sequence. But our prod DB already many of the migrations run when it was PG 9. So propose_migrations does not reflect what is actually deployed. It has been that way and I don't think there's anything we can do about it.

With all the changes, I still get 3 propose_migrations errors, which I just decided to suppress. I spent some time trying to fix the notifications ones with no luck, so I gave up. Here they are:

```
    <changeSet author="charlesoverbeck (generated)" id="1618014336064-3">
        <dropPrimaryKey constraintName="cloudinstance_supportedlanguages_pkey" tableName="cloudinstance_supportedlanguages"/>
    </changeSet>
    <changeSet author="charlesoverbeck (generated)" id="1618014336064-1">
        <modifyDataType columnName="id" newDataType="bigint" tableName="notification"/>
    </changeSet>
    <changeSet author="charlesoverbeck (generated)" id="1618014336064-2">
        <dropDefaultValue columnDataType="bigint" columnName="id" tableName="notification"/>
    </changeSet>
</databaseChangeLog>
+ grep changeSet dockstore-webservice/target/detected-migrations.xml
    <changeSet author="charlesoverbeck (generated)" id="1618014336064-3">
    </changeSet>
    <changeSet author="charlesoverbeck (generated)" id="1618014336064-1">
    </changeSet>
    <changeSet author="charlesoverbeck (generated)" id="1618014336064-2">
    </changeSet>
```

### CODE REVIEW

File: dockstore-webservice/pom.xml
Comment (wshands) [inline dockstore-webservice/pom.xml:942]: Any special reason to move to an older version of jaxb-runtime (2.3.3->2.3.1)?

**Comment by denis-yuen** [comment]
> Seems to work... I hope there wasn't some reason we were avoiding a 4.x Liquibase upgrade

Mentioned this previously, but just in case. Dropwizard is still on 3.x for the stable branch we're using, ex: https://github.com/dropwizard/dropwizard/releases/tag/v2.0.21

They are experimenting with it in their unstable branch, so potential beware

**Comment by denis-yuen** [comment]
Also "Upgarde" sounds cool

File: dockstore-webservice/src/main/resources/migrations.1.10.0.xml
Comment (denis-yuen) [inline dockstore-webservice/src/main/resources/migrations.1.10.0.xml:22]: Don't know how I feel about this, feels like historical revisionism

**Comment by coverbeck** [comment]
I'm getting reviews before I requested them! :) I wasn't quite ready.

File: dockstore-webservice/src/main/resources/migrations.1.10.0.xml
Comment (coverbeck) [reply dockstore-webservice/src/main/resources/migrations.1.10.0.xml]: Rolled it back, slipped in as an experiment.

File: dockstore-webservice/pom.xml
Comment (coverbeck) [reply dockstore-webservice/pom.xml:942]: > Any special reason to move to an older version of jaxb-runtime (2.3.3->2.3.1)?

The enforcer plugin was causing a failure after I upgraded the Hibernate version.

**Comment by coverbeck** [comment]
> > Seems to work... I hope there wasn't some reason we were avoiding a 4.x Liquibase upgrade
> 
> Mentioned this previously, but just in case. Dropwizard is still on 3.x for the stable branch we're using, ex: https://github.com/dropwizard/dropwizard/releases/tag/v2.0.21
> 
> They are experimenting with it in their unstable branch, so potential beware

I'm inclined to stick with the stable branch, at least for this iteration, to be on the safe side. But I can give it a shot if you think it's worth it.

**Comment by denis-yuen** [comment]
> I'm inclined to stick with the stable branch, at least for this iteration, to be on the safe side. But I can give it a shot if you think it's worth it.

I think probably not, unless it actually does fix the liquibase:diff problem.

File: .circleci/config.yml
Comment (denis-yuen) [inline .circleci/config.yml:151]: This is up from a default of medium?
Curious, why is this changing?

**Review by denis-yuen — APPROVED** [review]
> With all the changes, I still get 3 propose_migrations errors, which I just decided to suppress. I spent some time trying to fix the notifications ones with no luck, so I gave up. 

Fair enough

File: .circleci/config.yml
Comment (garyluu) [reply .circleci/config.yml:151]: There was a Nextflow test failing with 137

**Review by garyluu — APPROVED** [review]

**Review by GFJHogue — APPROVED** [review]

File: .circleci/config.yml
Comment (denis-yuen) [reply .circleci/config.yml:151]: Thinking the tests (or at least the slow ones) in this section should also be moved out to a job not called unit tests

**Review by denis-yuen — APPROVED** [review]

File: .circleci/config.yml
Comment (coverbeck) [reply .circleci/config.yml:151]: Yes, they've been failing intermittently on other branches too, without my changes, including [develop](https://app.circleci.com/pipelines/github/dockstore/dockstore/5763/workflows/2f485201-3ffd-4c53-bf2c-c771c1adf912/jobs/9860), so I upped it.

I'll take a quick look and see if there are obvious candidates to be split out.

**Review by NatalieEO — APPROVED** [review]

**Review by david4096 — APPROVED** [review]

**Comment by coverbeck** [comment]
Took a closer look at OOM errors on CircleCI jobs, and the first two I looked at were creating the Scala bridge compiler.

Turns out we have the Scala plugin in dockstore-integration-testing, but no Scala files there, so I removed it and dropped the memory back down.

Going to request re-review from Denis and Gary in case there's some reason I'm missing.

**Review by denis-yuen — APPROVED** [review]
> Turns out we have the Scala plugin in dockstore-integration-testing, but no Scala files there, so I removed it and dropped the memory back down.

> Going to request re-review from Denis and Gary in case there's some reason I'm missing.

If it works (tests), it works. 
This might have been a mis-understanding since the integration tests rely on the other modules which do use the Scala plugin

**Review by garyluu — APPROVED** [review]
A scala plugin has been in the dockstore-integration pom.xml for at least 3 years, A little surprisingly that the new one would be using notice-ably more memory.

**Comment by coverbeck** [comment]
> A scala plugin has been in the dockstore-integration pom.xml for at least 3 years, A little surprisingly that the new one would be using notice-ably more memory.

My guess is that we must have been close to the limit and something caused us to tip over it.

**Review by coverbeck — COMMENTED** [review]
Going to request one more round of reviews.

1. Got rid of Travis job and moved it into CircleCI
2. Renamed `unit-test` job to `unit_non-confidential_cwlrunner_tests`, which... But the unit tests were already running non-confidential tests, and I think part of the confusion is that the actual Java unit test step is fairly fast (6 minutes), but the the CircleCI job `unit_test` job took longer because it does more than unit tests.

Because of the rename and removal, some checks are failing, because they are no longer present. After the PR is merged, I'll change the checks.

**Review by garyluu — APPROVED** [review]

**Review by wshands — APPROVED** [review]

File: .circleci/config.yml
Comment (denis-yuen) [inline .circleci/config.yml:46]: Hmmm, miscommunication I think. I had in mind two possibilities.
Using the image as a guide.

Screenshot from 2021-04-13 11-58-18

a) Splitting out the non-confidential_cwlrunner tests into a fourth job on the right (I dislike having the unit tests combined with non unit tests) Additionally, the final stuff like building and publishing the Docker container seems unrelated too and should probably be dependent on all four jobs on the right anyway. Maybe they should be a final job. 
b) Moving the actual unit tests to the build step on the left and just leaving the cwlrunner tests here 

That said, this has absolutely nothing to do with hibernate or pg, so can just be new ticket rather than holding up this ticket.

**Review by denis-yuen — APPROVED** [review]

File: .circleci/config.yml
Comment (coverbeck) [reply .circleci/config.yml:46]: I'd thought of that, but I assumed there's overhead to additional jobs, and I was trying to avoid that.

I'll go ahead and rename the step back to unit_tests, to avoid less confusion and need for changing checks, then create a ticket for separation.

Comment (denis-yuen) [reply .circleci/config.yml]: I think we can live with the overhead since we're no longer on Travis-CI. 
Besides, running unit tests in the build step might rule out some obvious failures and skip the four on the right.

**Pull request merged**
